src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -478,46 +482,77 @@ private void InitializePackagesFolderInfo(string packagesFolderPath, bool exclud
                 throw new ArgumentNullException(nameof(secondarySources));
             }
 
-            if (nuGetProject is INuGetIntegratedProject)
+            var maxTasks = Environment.ProcessorCount;
+            var tasks = new List<Task<IEnumerable<NuGetProjectAction>>>(maxTasks);
+            var nugetActions = new List<NuGetProjectAction>();
+            
+            var buildIntegratedProjects = nuGetProjects.OfType<BuildIntegratedNuGetProject>().ToList();
+            var nonBuildIntegratedProjects = nuGetProjects.Except(buildIntegratedProjects).ToList();
+
+            // add tasks for all build integrated projects
+            foreach (var project in buildIntegratedProjects)
             {
+                // if tasks count reachs max then wait until an existing task is completed
+                if (tasks.Count == maxTasks)
```

Comment:
I confirmed this update operation for build integrated projects with different number of threads and 4 threads gave optimum result with cpu n memory. Even 4 threads had similar performance that with 8 threads for the update. So I'll go ahead with 4 threads here.